doc/PROTOCOL-SEMANTICS.md

```diff
@@ -0,0 +1,268 @@
+# gRPC semantics
+
+gRPC has many features, but the most fundamental core is the ability to perform
+RPCs. The goal of this document is defining the semantics of gRPC's RPCs.
+
+## Channels
+
+Most implementations will have Channel and Server concepts. A Channel is a
+virtual connection to an endpoint, capable of sending RPCs. Channel is "virtual"
+because the Channel is free to have zero or many actual connections. A Channel
+is free to determine which actual endpoint to use and may change it every RPC,
+permitting client-side load balancing. A Server is capable of receiving incoming
+connections and receiving RPCs.
+
+A "connection" is not a gRPC semantic concept and thus users should not assume a
+correlation between connections and RPCs. Although, practically, users should be
+aware of the common restriction that RPCs are unable to survive longer than the
+connection on which they exist.
+
+The exact details of how the communication is performed and higher level
+abstractions can change when necessary and are thus not a focus for this
+document. Although implementations SHOULD support the [HTTP/2
+transport](PROTOCOL-HTTP2.md) to provide a basis for interoperability.
+
+## Methods
+
+An RPC is performed on a Method. The Method has a name, defines the intended
+operation, the message types involved with the RPC, and the cardinality of
+those messages.  It does not define the endpoint to send the RPC to; this is
+normally handled by the Channel.
+
+A Method's intended operation is just normal documentation describing what a
+Method does, intended for a developer. gRPC itself is not generally aware of the
+intended operation.
+
+A Method has a request message type and a separate response message type. gRPC
+is only aware of these types well enough to serialize and deserialize them.
+Messages are considered opaque byte sequences of a known length to gRPC itself.
```

Comment:
It is important that messages are: 1) opaque and 2) can be serialized into byte sequences. Yes, a transport may be able to avoid the serialization, but in most languages inprocess still requires serializing and deserializing since messages may be mutable. And actually, in Java the marshaller is the thing aware that the serialization is unnecessary for inprocess, not gRPC itself.